### ISSUE TITLE
Help for bot, getting call notifications

### ISSUE THREAD

**Issue by stefan2410** [microsoftgraph/microsoft-graph-comms-samples#99]
Hi,
I'm trying to register/develop a bot (as a service), getting incoming calls notifications from incoming calls  and answer/transfer calls.  I'm using as a basic paradigm  the IncidentBot (trying to debug it locally with ngrok reserved domain) . I've registered the bot in Azure with the required API permissions and consent, then the Webhook  https://[ourdomain].ngrok.io/callback/calling
But I don't get incoming calls notifications. We want to relate the bot with a phone number to get incoming calls or (if no possible) with a group where the bot is the only one member.
It seems to me that I've not understood the workflow.My testing is  (as a teams user) to call the phone number  assigned to this group,  or to call another user. I expected, in any case, the bot to get all incoming calls notifications at least, but I'm not succesful. I don't want the bot to have interface as a user in teams. 
I would really appreciate your help. 
 Thanks

**Comment by stefan2410** [comment]
My webhook https://[service url]/callback/calling (this is the messaging endpoint in both MSTeams and Skype channels) has never been called from the Teams service. (using either ngrok or deployed AppService)
In my code, I've got the bot with DI in my Startup Configure and I can read the messages from CommunicationClient. It  seems normal, so the problem should be with the webhook only.

       CommunicationsClientBuilder.cs:122 Build  Assembly: GraphCommunicationsClient
        [INF] CommunicationsClientBuilder.cs:125 Build  Application: IncomingCallsBot
        [INF] CommunicationsClientBuilder.cs:128 Build  Product: GraphCommunicationsClient-IncomingCallsBot
        [INF] CommunicationsClientBuilder.cs:131 Build  Version: 1.1.0.581
        [INF] CommunicationsClientBuilder.cs:135 Build  UserAgent: GraphCommunicationsClient-IncomingCallsBot/1.1.0.581
        [INF] DispatcherContainer.cs:111 RegisterDispatcherGetOrAdd  Dispatcher Container registering notification dispatcher for: app/calls
        [INF] NotificationDispatcher.cs:225 Subscribe  Subscribed: app/calls To queue: app/calls

**Comment by stefan2410** [comment]
Please,  is it possible a bot to get incoming calls notifications without direct calling from Teams user interface, notifications for one group and possibly to answer them and/or transfer them, acccoringg to our rules ?
I would appreciate your answer.

**Comment by ksikorsk** [comment]
The only supported way to get incoming call notifications today is to call your bot using Teams client.  PSTN support is on it's way.

The answer/transfer APIs will be the same once PSTN support is added.  The only difference is that the source identity will be a phone identity.

**Comment by stefan2410** [comment]
Thanks @ksikorsk   but it was not clear in the documentation. It says only that a bot can't make outbound calls to PSTN.  When to expect the PSTN support?  a road maybe.

**Comment by rugt0r** [comment]
This is proving to be an issue for me- I was hoping that I would be able to use the graph API to get details of all ongoing calls, not just the ones involving the bot. I am trying to improve incoming call tracking, but including the bot in every call is an unwieldy solution that doesn't give us a complete picture of what's going on.

@ksikorsk do you know if there are any plans for an API that could expose all ongoing calls for a given tenant? If not, is there a means to request this? It would be more useful than the bot functionality.

Thanks.

**Comment by ksikorsk** [comment]
If I understand you correctly, you would like to be an observer of the meeting without actually being a participant?

@VinodRavichandran to track this request.

**Comment by VinodRavichandran** [comment]
Yes, Webhook subscription and notifications for Call events is something we have started designing. 

FYI @zhengni-msft

**Comment by rugt0r** [comment]
Thanks for the feedback, much appreciated.

**Comment by stefan2410** [comment]
@VinodRavichandran 
 Would be able to use the  "The Webhook subscription and notifications for Call events"  in an standalone application and in tenant scope?  or only in a bot?
There are 2 different use-cases
1. Get real-time call events (from incoming =>ringing =>answered =>hangup),  for all tenant users, by  a standalone application or a bot . The notifications will be used to update our history-analytics and for integrations.
2.  Get real-time call events  for a tenant's group (or groups) in a  bot . The bot to be able to answer the call (without the bot to be called directly)  play out IVR ( or extend the call queue IVR) , ring the appropriate users and transfer the call to first user answered it, get notifications for hangup.  We need also Presence availability to ring the users that we don't have this very moment.
Such functionality is part of this [user's voice issue](https://microsoftteams.uservoice.com/forums/555103-public/suggestions/38534776-full-teams-client-api-including-call-control) 
3. Can we hope the Webhook to be ready till MSIgnite (at least the 1st use-case) ?

**Comment by AHComp** [comment]
so 3 months later what is the progress in the desiging and implementation of notifications for Call events ?

**Comment by surezg** [comment]
@VinodRavichandran  - Looks like Teams Voice platform rollout was preponed to this december which was Q1-2020 before. Do we have any specific updates on the Bot Calling API's this month. 
Any tentative for full call recording and accepting PSTN to Bot which are not part of current API's.

Appreciate your help!

Thanks,
Suresh.

**Comment by VinodRavichandran** [comment]
Comms APIs were just released in v1 last week. PSTN number assignment is referred in the blog. Recording was not a part of the release

https://developer.microsoft.com/en-us/graph/blogs/microsoft-graph-cloud-communications-apis-are-now-generally-available/

**Comment by rugt0r** [comment]
> Yes, Webhook subscription and notifications for Call events is something we have started designing.
Hi, any update on when this is likely to be available? Being able to list all ongoing calls and their participants (without a bot having to be present in each call) would be very useful to us.

**Comment by Githamza** [comment]
Hello , I'm strongly interested too, I would like to keep track of notifications from an incoming call that is transfered to a user but without including the bot in the call/meeting ( user should see the incoming number/user when he respond to a call , not the Bot name ) .

Any updates , or workarounds ?

**Comment by scorpionfly7** [comment]
Hi, Is there ability to monitor calls in Teams, using Teams Client SDK? For example notify that there is started calls? @ksikorsk

**Comment by vipulw2011** [comment]
Hi, Is there any update on this issue. Even I need the notification of incoming call with it's details in my service via a callback url. If anyone has implemented can you please share/guide how we do it ? Thanks

**Comment by VishalWaman** [comment]
+1

**Comment by robert-leafworks** [comment]
**+1** from our side! 
For us we need the event for incoming phone call via PSTN as well. 
We need to **fire a webhook for phone call events** (i.e. ringing, call accepted) or building an app inside teams which opens when an incoming call is received. Both ways would work for us to implement our use cases.

**Comment by cyril-iselin** [comment]
+1 same like @robert-leafworks

**Comment by zindont** [comment]
+1 same like @robert-leafworks

**Comment by AHComp** [comment]
So guys now honestly Microsoft is just kidding you. There will definitely be no WebHook for the telephone functions because there is already a solution from Microsoft to implement such a function. Unfortunately, Microsoft does not want the smaller partner to develop solutions on this function because Microsoft wants to develop these solutions first with large software providers.
We noticed this over a year ago and have developed an alternative to the old TAPI interface based on the resources made available by Microsoft.
If somebody interested to buy the code or consulting it's possible to contact me.

**Comment by logistic-people-deutschland** [comment]
I am also interested
we want to build a solution
where we show our users which customer is calling
the customer data is in our database
whether there is something like that?

**Comment by babunirwan** [comment]
Any update on the webhooks for incoming PSTN calls notifications??

**Comment by robert-leafworks** [comment]
Maybe they will have in 2030 🤣  - until then we will discourage all clients to use MS Teams business voice for professional use - better use other solutions here.

How much effort would it be to simply have a webhook ready to be called when a call comes in? 
It's ridiculous that MS product managers didn't see this when building V1 of the application. It's like no one needs to integrate their phone channels with their ERP, CRM etc. 🙄

**Comment by babunirwan** [comment]
If they do not have incoming call hook where is this API used -
https://docs.microsoft.com/en-us/graph/api/call-answer?view=graph-rest-1.0&tabs=http

**Comment by siseli** [comment]
We need a call notification API for use in teams apps too.
Is it really the way it seems and this is still not possible?

The requirement is really simple: a web hook or some other event notification to monitor incoming and outgoing calls for a certain user.
We need it to display additional information about the call to the user.

Please MS, tell us that there is a way to do this or that you are at least working on it.

**Comment by qatestolegmig** [comment]
Any updates in 2024? Or MS does not care at all that customers request this feature for 5+ years?

**Comment by redek91** [comment]
I also need this feature, is it even possible to have a bot or webapp respond to a PSTN call in realtime? Even with the graph api subscription there is too much latency on a call created event, the bot has no time to pick up the call.

Is there any way to have the notification in realtime?

**Comment by robert-leafworks** [comment]
I also try to check this topic once a year to see if Microsoft was able to make this possible 😅.

Did anybody try this here already? 
It seems to be possible now to open URLs using Teams Call Policies --> "Open Apps in Browser for incoming PSTN calls" [Check this](https://learn.microsoft.com/en-us/microsoftteams/inbound-call-routing#open-apps-in-browser-for-incoming-pstn-calls)

In order to integrate with CRMs this could potentially help. Of course it's not a webhook but the moment when this occurs is the right one :)
